man/sysvinit.8
.TH ninit-sysvinit 8 "Dec 28, 2009"
.SH NAME
ninit-sysvinit \- SysVinit emulator listening on /dev/initctl
.SH SYNOPSIS
.B ninit-sysvinit
.br
.B ninit-sysvinit
powerS

.SH DESCRIPTION
.B ninit
does not use the
.I /dev/initctl
fifo.  As a result the SysVinit commands
telinit, reboot, halt, shutdown do not work.
Some of them need to write to 
.I /dev/initctl
to work properly.

.B ninit-sysvinit
is a SysVinit emulator which listens on
.I /dev/initctl
and starts some services if there is any activity on the fifo.

.SH USAGE
Five seconds after starting all services
.B ninit
tries to start the sysvinit service.
If the last service exists then 
.B ninit
opens and listens on
.I /dev/initctl. 
.B  Never include 
the service sysvinit in 
.I /etc/ninit/default
nor start it at boot time.

.SH OPTIONS
.B ninit\-sysvinit
reads the control file
.I sysvinit\-timeout
in the service directory.
It must contain two numbers n1:n2.
If there is not
any activity  on 
.I /dev/initctrl  
for more than n1 seconds
the program exits.
If the number n1 is zero
.B ninit\-sysvinit
wait forever.

If needed, 
.B ninit
can start
.B ninit\-sysvinit
later again.

.SH RUNLEVELS
SysVinit has runlevels 0123456S.  The command
.br
.B telinit S
.br
writes the request to
.I /dev/initctl
and 
.B ninit\-sysvinit
starts the service 
.I levelS.
It works similarly with telinit 5, telinit Q or telinit U.
Then
.B ninit\-sysvinit
starts the 
.I level5,
.I levelQ 
or 
.I levelU 
service if it exists.

.SH "POWER ACTIONS"
.B ninit\-sysvinit
also starts the services
.I powerF,
.I powerO 
or
.I powerL.
The letters come from F(AIL), O(K) or L(OW).  
The program can obtain the power request from
.I /dev/initctl
or from a SIGPWR signal and the file
.I /etc/powerstatus.
See the manual page init(8) for more info.

.SH SIGNALS
Don't send signal SIGPWR to
.B ninit\-sysvinit.
Manpage init(8) says:


Usage of SIGPWR and 
.I /etc/powerstatus 
is discouraged. Someone wanting to
interact  with  init  should use the 
.I /dev/initctl 
control channel.

If some program still needs to use signal SIGPWR
it must send it dirrectly to
.B ninit.

.SH ENVIRON
SysVinit 
.B shutdown
sets the variable 
INIT_HALT=POWERDOWN 
or
INIT_HALT=HALT 
depending on the options
\-r \-h \-H \-P.  You can (re)set this variable
using:

.B nsvc -E INIT_HALT
.br
.B nsvc -E INIT_HALT=POWERDOWN

.SH "AUTHOR"
If you don't plan to use SysvInit's shutdown, reboot,
halt or telninit commands, you can do:

.B cd /etc/ninit
.br
.B mv sysvinit sysvinit_
.br
.B ninit\-reload \-v \-u /sbin/ninit

Why did I write this SysVinit 
.B emulator?  
This is a good question.
Ninit has the ninit-shutdown, nkillall and ninit-reboot programs.  There
is no need to use SysVinit halt, reboot... to stop the box clean.
See the code of the emulator.  Maybe you will find the answer there.

Nikola Vladov

.SH FILES
.I /dev/initctl
.br
.I /etc/powerstatus
.br
.I NINIT_HOME/sysvinit/sysvinit\-timeout

.SH SEE ALSO
nsvc(8), ninit-shutdown(8), ninit-runlevel(8)
.br
init(8), shutdown(8), halt(8), reboot(8), telinit(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


sysvinit.c
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <sys/poll.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <alloca.h>
#include "ninitfeatures.h"
#include "initreq.h"
#define IREQ sizeof(struct init_request)

static char read_buf[IREQ + 4];
struct init_request *req = (void *)read_buf;
static int infd,outfd,flag_fork;

#include "open_inout.h"
#include "uid.h"

void sighandler(int sig) { (void)sig; while (waitpid(-1,0,WNOHANG) > 0); }

static void do_it(char *A) {
  int r=0;
  if (flag_fork) { while ((r=fork()) == -1) nano_sleep(1,0); }
  if (r==0) {
    open_inout(INITROOT);
    msg("Starting service: ",A+1);
    write(infd, A, str_len(A));
    read(outfd, req, IREQ);
    
    close(infd); close(outfd);
    if (flag_fork) _exit(0);
  }
}

void print_env() {
  char *x = req->i.data;
  msg("request"," for INIT_CMD_SETENV");
  while (*x) {
    msg(x);
    x += str_len(x);
    x++;
  }
}

int main(int argc, char **argv) {
  struct pollfd pfd;
  unsigned long ul[2] = {0,0};
  int fd, timeout = -1;
  char power[] = "spowerS";

  errmsg_iam("ninit-sysvinit");
  
  if (argc>1 && !str_diff(argv[1], "powerS")) {
    char ch;
    msg("trying to read and remove: ", INIT_POWERSTATUS);

    fd=open(INIT_POWERSTATUS, O_RDONLY);
    if (fd>=0) { read(fd,&ch,1); close(fd); }
    unlink(INIT_POWERSTATUS);
    
    if ('a' <= ch && ch <= 'z') ch -= 32;
    switch (ch) {
      case 'L': case 'O': break;
      default: ch = 'F';
    }
    
    power[6] = ch;
    do_it(power);
    return 0;
  }

  fd = open(INIT_FIFO, O_RDWR | O_NONBLOCK);
  if (fd < 0) {
    SYS_mknod(INIT_FIFO,S_IFIFO|0600,0);
    return -1;
  }
  fcntl(fd,F_SETFD,FD_CLOEXEC);

  pfd.fd = fd;
  pfd.events = POLLIN;
  read_ulongs("sysvinit-timeout", ul, 2);

  if (ul[0] > 0)timeout = 1000 * ul[0];
  flag_fork = (ul[1] != 0);
  // msg(" timeout: ",fu(ul[0]), ",fork-mode: ",flag_fork ? "YES" : "NO");

  if (flag_fork) set_sa(SIGCHLD);

  while (1) {
    int flag_known = 0;
    switch (poll(&pfd,1,timeout)) {
    case -1:
      if (errno == EINTR) sighandler(SIGCHLD); break;
      return -1;	/* poll failed */
    case 1: {
      int r = read(fd, req, IREQ);
      if (r == (int)IREQ) {

#ifdef SYSVINIT_DUMP_INITREQ
	if ((r=open("/tmp/__##initreq", O_WRONLY | O_APPEND)) >= 0) 
	  {  write(r,req,IREQ);  close(r); }
#endif

	if (req->magic == INIT_MAGIC) {
	  char level[] = "slevelS";
	  char *A="";

	  switch (req->cmd) {
	  case INIT_CMD_SETENV:
	    print_env();
	    flag_known=1;
	    break;
	  case INIT_CMD_RUNLVL:
	    r = req->runlevel;
	    if ('a' <= r && r <= 'z') r -= 32;
	    level[6] = r;
	    A = level;
	    break;
	  case INIT_CMD_POWEROK:
	    A = power;
	    power[6] = 'O';
	    break;
	  case INIT_CMD_POWERFAIL:
	    A = power;
	    power[6] = 'F';
	    break;
	  case INIT_CMD_POWERFAILNOW:
	    A = power;
	    power[6] = 'L';
	  }

	  if (*A) { flag_known=1; do_it(A); }
	}	/* INIT_MAGIC	*/
      }		/* r == sizeof(...) */
      if (flag_known==0) carp("unknown", " request");
    }
    break;
    default:	/* timeout	*/
      return 0;
    }		/* poll		*/
  }
}
